2주차 Step2 - 연관 관계 매핑

devsigner9920 opened the pull request:
안녕하세요!

연관 관계 매핑 해보았습니다.

영한님 책 보며 하나하나 차근차근 해보았습니다!

한가지 질문 사항이 있습니다!

설정 관련해서 ddl create-auto 옵션을 주었는데 테스트 시에 외래키 제약조건이 생성되어 있질 않아 처음에 외래키 조건을 드롭하는 프로세스가 돌아갈 때 내부적으로 계속 오류를 내더라구요.. 테스트 케이스 수행에 문제는 안되지만 찝찝해서 계속 찾아봤는데 해결을 못했습니다 ㅠ

혹시 해결 방법을 찾을 수 있을까요..?!

매번 리뷰 감사드립니다! 잘부탁드립니다!

devsigner9920 committed:
docs(README): 2주차 1단계 미션을 위한 요구사항 명세 및 TODO 리스트 작성

devsigner9920 committed:
feat(schema, data): H2 데이터베이스에서 인식할 스키마 및 데이터 파일 생성

devsigner9920 committed:
feat(application.properties): JPA 설정

devsigner9920 committed:
feat(BaseTimeEntity): 생성일, 수정일을 자동으로 세팅하는 클래스 구현

추가 내용:
- 엔티티들이 공통으로 가지는 생성일, 수정일 컬럼을 엔티티에 매번 작성하지 않고 공통으로 관리하기 위해 생성하였습니다.

devsigner9920 committed:
refactor(Answer): JPA Entity 객체로써 인식할 수 있도록 변경

devsigner9920 committed:
style(convention): 4 space 인덴트를 2space로 변경하였습니다.

devsigner9920 committed:
feat: 도메인 모델들을 JPA 엔티티로 변경

devsigner9920 committed:
fix: 어플리케이션 설정 변경 및 데이터 파일 삭제, 스키마 MySQL 문법으로 변경

devsigner9920 committed:
refactor: 도메인 객체들을 엔티티 형태로 리팩토링

devsigner9920 committed:
feat: 도메인별 저장소 JPA 학습 테스트를 위한 메소드 추가

devsigner9920 committed:
feat: JPA 학습 테스트 실행

devsigner9920 committed:
refactor: 엔티티 객체 가독성 향상 및 타입 명시, 불필요한 인덱스 제거

devsigner9920 committed:
refactor: @Param 어노테이션을 적용하여 update문 개선

devsigner9920 committed:
test: 단위 테스트 진행 시 누락된 데이터 초기화 부분 개선 및 전체 테스트 통과 완료

devsigner9920 committed:
refactor: 요구사항 미반영 수정 완료

devsigner9920 committed:
refactor: 프로퍼티 설정 부분을 DataJpaTest 진행을 위하여 변경

devsigner9920 committed:
docs: 2단계 미션 수행을 위한 요구사항 및 To do 리스트 정리

devsigner9920 committed:
refactor: 테스트를 위한 환경 설정 변경

devsigner9920 committed:
refactor: User와 질문, 답변의 연관관계 매핑

추가 내용:
- 사용자 한명에 여러개의 질문이 있을 수 있으므로 OneToMany 관계로 매핑했습니다.
- 사용자 한명에 여러개의 답변이 있을 수 있으므로 OneToMany 관계로 매핑했습니다.
- 질문과 답변 모두 사용자가 삭제될 경우 작성자가 없어져야 한다고 생각하므로 cascade 옵션을 주었습니다.

devsigner9920 committed:
refactor: 질문에 대한 사용자, 답변들에 대한 연관관계 매핑

추가 내용:
- 각 질문에 대한 작성자는 한명이므로 ManyToOne 연관 관계를 맺어주었습니다.
- 하나의 질문에 여러 답변이 있을 수 있으므로 OneToMany 연관 관계를 맺어주었습니다.

devsigner9920 committed:
refactor: Question Entity 변경에 따른 코드 일부 수정

devsigner9920 committed:
refactor: Answer 객체와 writer, question에 대한 연관관계 매핑

반영내용:
- 각 Answer은 하나의 답변자와 질문을 가지기 때문에 ManyToOne의 관계를 맺었습니다.

devsigner9920 committed:
refactor: DeleteHistory와 User 간의 연관관계 매핑

반영내용:
- 하나의 사용자가 여러 삭제 히스토리를 가지고 있을 수 있다고 판단했습니다.
- 처음에 1:1 관계를 적용하려 하다 하나의 유저가 다건의 삭제 내역을 가질 수 있겠다 생각하고 변경하였습니다.

devsigner9920 committed:
test: 엔티티 테스트를 위한 테스트용 엔티티팩토리 생성

devsigner9920 committed:
test: 연관관계 매핑 이후 각 테스트케이스 실패 발생 대처 및 테스트 케이스 리팩토링

devsigner9920 committed:
docs: git resolve conflict

devsigner9920 committed:
Merge branch 'devsigner9920' into step2

nokchax commented:
```
설정 관련해서 ddl create-auto 옵션을 주었는데 테스트 시에 외래키 제약조건이 생성되어 있질 않아 
처음에 외래키 조건을 드롭하는 프로세스가 돌아갈 때 내부적으로 계속 오류를 내더라구요..
```
create-auto 는 실행시에 drop을 먼저 실행하기 때문에 error는 어쩔수 없는것 같습니다. 에러가 보기 싫으시다면 update를 사용해보세요

라고 남기고 제가 테스트좀 해봤는데 오류가 많이 나네요ㅠ
일단 연관관계 설정에서 JoinColumn 때문에 매 테스트시에 sql에러가 납니다.
제가 수정했던 과정을 공유해 드릴게요

0. src/main/resources/data/schema.sql 삭제

1. Question 수정
```java
  @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL, mappedBy = "question")
  private final List<Answer> answers = new ArrayList<>();
```

2. User 내의 JoinColumn 어노테이션 모두 삭제
여기서 생성된 외래키 때문에 에러가 나더라구요

3. id autoincrement 초기화
이후에도 전체 테스트를 돌리면 에러가 나는데요, update로 변경하면서 db가 완전히 초기화 되지 않기 때문에 발생하는 문제입니다.
테스트 전반적으로 id를 사용한 조회가 많은데요, dirtycontext를 사용하여 초기화를 하더라도 autoincrement는 초기화 되지 않기 때문에
매 테스트가 끝난 후 초기화 하는 식으로 하시면 됩니다. (사실 id 값에 영향받지 않도록 테스트 하는게 좋은 방법 같습니다..)
```
// usertest내에 추가
  @BeforeEach
  void reset() {
      entityManager.createNativeQuery("ALTER TABLE `user` ALTER COLUMN `id` RESTART WITH 1")
              .executeUpdate();
  }
// questiontest내에 추가
  @BeforeEach
  void reset() {
    entityManager.createNativeQuery("ALTER TABLE `question` ALTER COLUMN `id` RESTART WITH 1")
            .executeUpdate();
  }
```

![image](https://user-images.githubusercontent.com/32362748/141674185-3f3387df-9c15-46e2-a0b4-119331851a10.png)

nokchax commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -12,11 +12,13 @@
+  private Question question;
 
-  @Column(columnDefinition = "longtext")
+  @Lob
```
👍

nokchax commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,97 +1,109 @@
-                ", deleted=" + deleted +
-                '}';
-    }
+  @Id
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private Long id;
```
indent 4로 통일해 주세요 ㅠ

devsigner9920 replied:
@nokchax 죄송합니다 ㅠ 컴퓨터를 바꾸면서 프로젝트 인덴트가 설정해둔게 초기화 됐나봐요! 다음엔 확실하게 확인하겠습니다!!

nokchax commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,97 +1,109 @@
+  @Column(columnDefinition = "bit", nullable = false)
+  private boolean deleted = false;
+
+  @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+  @JoinColumn(name = "question_id", foreignKey = @ForeignKey(name = "fk_answer_to_question"))
+  private final List<Answer> answers = new ArrayList<>();
```
`@JoinColumn` 과 `mappedBy` 에 대해서 찾아보세요 :)

devsigner9920 replied:
넵 찾아서 학습 해보았습니다!! 좋네요 ㅎㅎ 연관관계 매핑 좀 더 알아가는 것 같습니다 ㅎ

nokchax commented on src/main/java/qna/domain/User.java:
```diff
@@ -25,6 +27,18 @@
   @Column(nullable = false, length = 50)
   private String email;
 
+  @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+  @JoinColumn(name="writer_id", foreignKey = @ForeignKey(name = "fk_question_writer"))
+  private final List<Question> questions = new ArrayList<>();
+
+  @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+  @JoinColumn(name="writer_id", foreignKey = @ForeignKey(name = "fk_question_writer"))
+  private final List<Answer> answers = new ArrayList<>();
+
+  @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+  @JoinColumn(name = "deleted_by_id", foreignKey = @ForeignKey(name = "fk_delete_history_to_user"))
+  private final List<DeleteHistory> deleteHistories = new ArrayList<>();
```
전반적으로 양방향 연관관계 설정을 하셨는데요, 단방향에 비해 양방향은 신경쓸 것이 많아 진답니다.
현재 요구사항에서 양방향이 꼭 필요할까요?

devsigner9920 replied:
@nokchax 처음에 단방향 연관관계를 설정해보았는데요! 단순히 생각했을 땐 양방향으로 하면 User쪽에서 객체를 조회하는게 쉽지 않을까? 하는 의문으로 해보았습니다.

그런데 찾아보니 양방향 관계일 시 관리에 조금 더 신경을 써야하고, 무한 루프에 빠지지 않도록 조심해야하는 등 현재 단계에서 제가 신경 쓰지 못하는 점들이 많아 보입니다!

영한님 책을 보면서 조금 더 확실히 학습한 후에 필요한 경우에만 쓰도록 하겠습니다 !

nokchax commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,67 +1,74 @@
+  private Question q1;
+  private Question q2;
+
+  @BeforeEach
+  void setUp() {
+    q1 = QuestionFactory.create("test", "contents");
+    q2 = QuestionFactory.create("test q2", "contents q2");
```
setup과 테스트에 대해서 한번 생각해보셨으면 해서 공유드려요
https://jojoldu.tistory.com/611
https://blog.kingbbode.com/52

devsigner9920 replied:
아하... setup에서 선언하기 보다는 테스트 케이스 내부에서 하는게 좀 더 직관적이고 공유할 일이 없어지는 것 같네요!

좋은 내용 감사합니다!! 적용 해보겠습니다 ㅎㅎ

nokchax reviewed (changes requested):
안녕하세요, 원준님
연관 관계매핑 진행 고생많으셨습니다 👏👏
질문주신 것과 코멘트 남겼는데요, 확인 후 재요청 부탁드립니닫 :)

devsigner9920 committed:
fix: 테스트 수행 시, 에러 발생 해결

내용:
- ddl create-drop을 하면서 존재하지 않는 외래키가 삭제될 경우 에러가 발생하는 이슈가 있었습니다.
- 해당 이슈를 해결하기 위해 리뷰어님의 피드백을 받아 적용해보았습니다.

devsigner9920 committed:
refactor: 양방향 연관관계 설정을 단방향 연관관계로 수정

devsigner9920 commented:
@nokchax 아하.. 질문한 내용에 대한 세심한 답변 감사합니다!!
JPA랑 조금 더 친해져서 오류 없이 테스트를 할 수 있게끔 노력해보겠습니다!
정말 퀄리티 높은 답변을 해주셔서 감동이에요 ㅎㅎ
피드백 항상 감사드립니다!

nokchax commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -16,6 +16,7 @@
 
   @ManyToOne
+  @JoinColumn(name = "deleted_by_id", foreignKey = @ForeignKey(name = "fk_delete_history_to_user"))
   private User deletedById;
```
User 타입이므로 필드명 deletedById가 어울리지 않게 되었네요 :)

devsigner9920 replied:
아하 넵 deletedByUser로 해주는게 더 알맞는 명명 규칙이군요! 세세한 디테일 신경 쓰겠습니다

nokchax commented on src/main/java/qna/domain/Question.java:
```diff
@@ -16,16 +16,13 @@
-  @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
-  @JoinColumn(name = "question_id", foreignKey = @ForeignKey(name = "fk_answer_to_question"))
-  private final List<Answer> answers = new ArrayList<>();
-
```
Question 과 Answer 간의 관계에 대해서 한번 생각해 봐야할 것 같아요.
지금은 다대일 단방향 관계로 설정해 주셨는데요, 이러면서 Question 에서 Answer를 다루기 힘들게 됐습니다.

QnA 서비스를 생각해볼게요, 하나의 Answer에서 Question을 참조할 일이 많을까요?
아니면 하나의 Question에서 여러 Answer를 가져다 사용할 일이 많을까요?

Question과 Answer간의 관계뿐만 아니라 다른 엔티티 간의 관계도 한 번 되돌아 볼 필요가 있을거 같아요.
지난주 목요일에 진행했던 JPA 수업에서 소개된 여러 관계에 대해 깊이 생각해보시고, 도메인 간의 적절한 관계를 찾아보세요 :)

devsigner9920 replied:
아하 넵 알겠습니다 수업 다시 복습하고 완전히 슥듭하겠습니다! 그러네요.. 단순히 외래키 생성되는 곳을 주인격으로 생각하지 않고 조금 더 실 사용에서의 관계를 초점에 두고 해보겠습니다! 감사합니다 ㅎㅎ

nokchax commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -20,18 +22,23 @@
+      .executeUpdate();
+
+    Question q1 = QuestionFactory.create("test", "contents");
+    Question q2 = QuestionFactory.create("test q2", "contents q2");
```
이전 코멘트에도 남기긴 했지만, 보다 좋은 방식은 id값에 의존하지 않는 테스트를 짜는 거라고 봐요.
예를 들어볼게요 findById() 의 경우에
```java
...
questionRepository.save(q1);

Question question = questionRepository.findById(1L).orElse(null);
...
```
처럼 `Question question = questionRepository.findById(1L).orElse(null);` 1L을 사용하셨는데, 대신에
```java
...
savedQuestion = questionRepository.save(q1);

Question question = questionRepository.findById(savedQuestion.getId()).orElse(null);
...
```
`savedQuestion.getId()` 처럼 테스트를 수정하면, 매번 id값을 1로 초기화 할 필요도 없어지겠죠 :)
지난번 피드백에 이렇게 안내 드릴걸 그랬네요 ㅠ 이 피드백은 반영 안하셔도 됩니다. 더 좋은 방식을 안내해 드린것 뿐이에요

devsigner9920 replied:
앗 그래도 조금 더 확실하게 배워서 좋습니다!!! 감사합니다 테스트 케이스 그래도 한번 더 반영 해보겠습니다! ㅎㅎ

nokchax reviewed (changes requested):
안녕하세요, 원준님! 피드백 반영하느라 고생하셨습니다. 👏👏 
몇가지 추가적으로 피드백 드린 부분이 있는데, 확인후 재요청 부탁드립니다 💪

devsigner9920 committed:
refactor: 리뷰어님의 피드백 반영

- 연관관계에 대해서 조금 더 객체지향적으로 접근하기 편하게 개선하였습니다.
- 객체 그래프 탐색을 할 경우 어떤 구조가 좋을지 고민하며 피드백을 반영하였습니다.
- DeleteHistory에서 deletedById를 deletedBy로 변경하였습니다.
- 테스트 케이스에서 아이디를 지정하여 사용한 부분을 리팩토링 하였습니다.

nokchax reviewed (approved):
안녕하세요, 원준님 피드백 반영 고생하셨습니다. 👍 
머지할게요 다음 단계 진행해주세요 :)
